app/assets/javascripts/angular/controllers/graph_ctrl.js

```diff
@@ -143,7 +147,18 @@ angular.module("Prometheus.controllers").controller('GraphCtrl',
     $scope.addAxis();
   }
 
+  $scope.$watch('graph.range', $scope.refreshGraph);
+  $scope.$watch('graph.endTime', $scope.refreshGraph);
+  $scope.$watch('graph.resolution', $scope.refreshGraph);
+  $scope.graph.expressions.forEach(function(e) {
```

Comment:
This only covers expressions that are there at dashboard load, but not new ones... I guess the expression would have to fire a `setServer` event (or similar), like you do with the changed expression?